third_party/nvfuser/csrc/multidevice/aggregate_dag.h

```diff
@@ -0,0 +1,164 @@
+#pragma once
+#include <fusion.h>
+#include <ir_base_nodes.h>
+#include <ir_container.h>
+#include <ir_printer.h>
+#include <iter_visitor.h>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+class Group;
+class MultiGroupFusion;
+class AggregateVal;
+
+// TODO: Revisit
+using GroupPtr = std::shared_ptr<Group>;
+
+// Implements the AggregateDag. It is build from a multigroup fusion.
+// Its traversal is what defines the runtime execution.
+// The nodes of the dag are the Vals AggregateVal
+// The edges are Expr of either type 1) AggregateExpr which represent the
+// operations contained in a Group, or 2)SendRecv which represents the
+// collective operation between two processes that needs to be done at runtime.
+
+// AggregateDag is an IrContainer, built from a multigroup fusion.
+// Its traversal is what orders the runtime execution.
+class TORCH_CUDA_CU_API AggregateDag : public Fusion {
+ public:
+  AggregateDag();
+
+  // Build the AggregateDag from a MultiGroupFusion.
+  // Maybe we should eventually build the Aggregate dag as we go along with the
+  // Fusion, creating AggregateVal, AggregateExpr as soon as it is needed (e.g.
+  // when we call newGroup)
+  void build(MultiGroupFusion* fusion);
+
+  // prints the dag
+  void print();
+
+ private:
+  friend MultiGroupFusion;
+
+  // Store the producer of each Val of the DAG
+  std::unordered_map<Val*, AggregateVal*> producer;
+  // Store the consumers of each Val of the DAG
+  std::unordered_multimap<Val*, AggregateVal*> consumers;
+};
+
+
+// An AggregateVal is basically a Val and a Group. At runtime, since a Group is
+// typically associated with a rank, AggregateVal represents a Val which is
+// located at a rank.
+class TORCH_CUDA_CU_API AggregateVal : public Val {
+ public:
+  AggregateVal(IrBuilderPasskey passkey, Val* val, GroupPtr group);
+  AggregateVal(const AggregateVal* src, IrCloner* ir_cloner);
+
+  NVFUSER_DECLARE_CLONE
+
+  virtual std::string toString(int indent_size = 0) const override;
+  virtual std::string toInlineString(int indent_size = 0) const override;
+
+  Val* getOriginalVal() const {
+    return original_val_;
+  }
+
+  const GroupPtr getGroup() const {
+    return group_;
+  }
+
+  bool sameAs(const Statement* other) const override;
+
+ private:
+  Val* original_val_;
+  GroupPtr group_;
```

Comment:
Do we need more than just the rank of the val? Would like to decouple the MultiDevice and AggregateDag classes.

Reply:
I am not sure to understand. Is this comment still relevant after these last changes?